docs/design/sql/04-mulitthreaded-execution.md

```diff
@@ -0,0 +1,215 @@
+execution is initiated in response to query start or data messages. Fragment execution may take significant time to complete.
+
+Second, the engine must process query operations described in [[1]] (p. 3), such as query start, query cancel, batch arrival,
+and flow control. These operations take short time to complete. Moreover, query start operation may trigger the parallel
+execution of several fragments.
+
+Given the different nature of query operations and fragment execution, we split them into independent stages, called
+**operation pool** and **fragment pool**. The former executes query operations, and the latter executes fragments. This design
+provides a clear separation of concerns and allows us to optimize stages for their tasks as described below, which improves
+performance. On the other hand, this design introduces an additional thread notification, as shown in the snippet below, which
+may negatively affect performance. Nevertheless, we think that the advantages of this approach outweigh the disadvantages.
+
+*Snippet 2: Query start flow (receiver only)*
+```
+IO               Operation pool         Fragment pool
+|----enqueue/notify-->-|                      |
+|                      |----enqueue/notify-->-|
+```
+
+### 2.2 Operation Pool
+
+Every network message received by the IO pool is first submitted to the operation pool.
+
+The pool is organized as a fixed pool of threads with dedicated per-thread queues. Dedicated queues reduce contention and
+increase the throughput, which is important given that processing of a single message takes little time.
+
+Every message may define an optional logical partition. If the partition is not defined, a random thread is picked for message
+processing. If the partition is defined, the index of executing thread is defined as `threadIndex = partition % threadCount`.
+Two messages with the same partition are guaranteed to be processed by the same thread, thus providing ordering guarantees.
+
+Some messages may trigger query fragment execution, as described below. In this case, a fragment execution task is created and
+submitted for execution to the fragment pool.
+1. `execute` - triggers the execution of one or more fragments
+1. `batch` and `flow_control` - trigger execution of one fragment defined by query ID and edge ID, as described in [[1]]
+
+### 2.3 Fragment Pool
+
+The fragment pool is responsible for the execution of individual query fragments. Fragment execution is always initiated by
+a task from the operation pool.
+
+Unlike operations, fragment execution may take arbitrary time depending on the query structure. It is, therefore, important to
+guarantee high throughput for short fragments, while still providing load balancing for long fragments. The ideal candidate
+is a thread pool with dedicated per-thread queues and work-stealing. For this reason, we choose JDK's `ForkJoinPool` as a
```

Comment:
There is one potential issue with `ForkJoinPool`. If I understand its internals correctly, under the load and uneven task sizes the pool prefers new task submissions over the old ones. Consider a pool of two queues with a constant flow of new submissions, the thread associated with the first queue is busy with the execution of some big task, the second thread finishes with another task execution, inspects its local queue, there is always something in it, since we have a constant flow of submissions, the tasks submitted to the first queue (potentially long time ago) are postponed.

Reply:
We do not consider any kind of prioritization at the moment. The system is optimized for throughput, so if there is a constant flow of tasks, and the older task is executed after the newer ones, it is still ok, because it was postponed due to another work of equal priority.

If this will be the problem in the future, we may consider cooperative interrupts, where a task finishes its execution after some time.